File: examples/cblas/dgemm.c
// Copyright (C) 2026 Zynum contributors
// SPDX-License-Identifier: LGPL-3.0-or-later

#include <stdio.h>

#include <zynum/blas/cblas.h>

int main(void) {
    const int m = 2;
    const int n = 2;
    const int k = 3;

    /* This CBLAS example uses row-major storage. */
    const double a[] = {
        1.0, 2.0, 3.0,
        4.0, 5.0, 6.0,
    };
    const double b[] = {
        7.0,  8.0,
        9.0, 10.0,
        11.0, 12.0,
    };
    double c[4] = {0.0, 0.0, 0.0, 0.0};

    cblas_dgemm(
        CblasRowMajor,
        CblasNoTrans,
        CblasNoTrans,
        m,
        n,
        k,
        1.0,
        a,
        k,
        b,
        n,
        0.0,
        c,
        n
    );

    puts("C = A x B");
    for (int row = 0; row < m; row += 1) {
        for (int col = 0; col < n; col += 1) {
            printf("%8.1f", c[row * n + col]);
        }
        putchar('\n');
    }

    return 0;
}

File: examples/fortran/dgemm.f90
! Copyright (C) 2026 Zynum contributors
! SPDX-License-Identifier: LGPL-3.0-or-later

program zynum_fortran_dgemm_example
  use, intrinsic :: iso_c_binding, only: c_char, c_double
  use zynum_blas_fortran, only: blasint, dgemm
  implicit none

  integer(blasint), parameter :: m = 2
  integer(blasint), parameter :: n = 2
  integer(blasint), parameter :: k = 3
  real(c_double), parameter :: alpha = 1.0_c_double
  real(c_double), parameter :: beta = 0.0_c_double
  character(kind=c_char), parameter :: no_trans = 'N'

  ! Fortran and BLAS use column-major storage.
  real(c_double) :: a(m * k) = [ &
      1.0_c_double, 4.0_c_double, &
      2.0_c_double, 5.0_c_double, &
      3.0_c_double, 6.0_c_double &
  ]
  real(c_double) :: b(k * n) = [ &
      7.0_c_double, 9.0_c_double, 11.0_c_double, &
      8.0_c_double, 10.0_c_double, 12.0_c_double &
  ]
  real(c_double) :: c(m * n) = 0.0_c_double

  call dgemm(no_trans, no_trans, m, n, k, alpha, a, m, b, k, beta, c, m)

  print *, 'C = A x B'
  call print_column_major_matrix(m, n, c)

contains

  subroutine print_column_major_matrix(rows, cols, values)
    integer(blasint), intent(in) :: rows
    integer(blasint), intent(in) :: cols
    real(c_double), intent(in) :: values(*)
    integer(blasint) :: row
    integer(blasint) :: col

    do row = 1, rows
      write(*, '(*(f8.1))') (values(row + (col - 1) * rows), col = 1, cols)
    end do
  end subroutine print_column_major_matrix

end program zynum_fortran_dgemm_example
